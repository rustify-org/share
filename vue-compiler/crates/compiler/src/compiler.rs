use super::{
    SFCInfo,
    codegen::{CodeGenerateOption, CodeGenerator, CodeGen, ScriptMode, CodeGenInfo},
    converter::{
        no_op_directive_convert, BaseConvertInfo as BaseInfo, BaseConverter, BaseRoot,
        ConvertOption, Converter, DirConvertFn, V_BIND, V_MODEL,
    },
    error::{NoopErrorHandler, RcErrHandle},
    flags::RuntimeHelper,
    parser::{Element, ParseOption, Parser, WhitespaceStrategy, AstRoot},
    scanner::{ScanOption, Scanner, TextMode, Tokens},
    transformer::{BaseTransformer, CorePass, TransformOption, Transformer},
    util::{no, yes},
    Namespace,
    transformer::{
        collect_entities::EntityCollector,
        mark_patch_flag::PatchFlagMarker,
        mark_slot_flag::SlotFlagMarker,
        optimize_text::TextOptimizer,
        pass::{Scope, SharedInfoPasses},
        process_expression::ExpressionProcessor,
        hoist_static::HoistStatic,
    },
};

use rustc_hash::FxHashMap;
use std::{io, rc::Rc, marker::PhantomData};

pub struct CompileOption {
    /// e.g. platform native elements, e.g. `<div>` for browsers
    pub is_native_tag: fn(&str) -> bool,

    /// e.g. native elements that can self-close, e.g. `<img>`, `<br>`, `<hr>`
    pub is_void_tag: fn(&str) -> bool,

    /// e.g. elements that should preserve whitespace inside, e.g. `<pre>`
    pub is_pre_tag: fn(&str) -> bool,

    /// Platform-specific built-in components e.g. `<Transition>`
    /// The pairing runtime provides additional built-in elements,
    /// Platform developer can use this to mark them as built-in
    /// so the compiler will generate component vnodes for them.
    pub get_builtin_component: fn(&str) -> Option<RuntimeHelper>,

    /// Separate option for end users to extend the native elements list
    pub is_custom_element: fn(&str) -> bool,

    /// Get tag namespace
    pub get_namespace: fn(&str, Option<&Element<'_>>) -> Namespace,

    /// Get text parsing mode for this element
    pub get_text_mode: fn(&str) -> TextMode,

    /// @default ['{{', '}}']
    pub delimiters: (String, String),

    /// Whitespace handling strategy
    pub whitespace: WhitespaceStrategy,

    /// platform speicific helper
    pub helper_strs: &'static [&'static str],

    /// Whether to keep comments in the templates AST.
    /// This defaults to `true` in development and `false` in production builds.
    pub preserve_comments: Option<bool>,
    /// Whether the output is dev build which includes v-if comment and dev patch flags.
    pub is_dev: bool,

    /// An object of { name: transform } to be applied to every directive attribute
    /// node found on element nodes.
    pub directive_converters: FxHashMap<&'static str, DirConvertFn>,
    /// Hoist static VNodes and props objects to `_hoisted_x` constants
    /// @default false
    pub hoist_static: bool,
    /// Cache v-on handlers to avoid creating new inline functions on each render,
    /// also avoids the need for dynamically patching the handlers by wrapping it.
    /// e.g `@click="foo"` by default is compiled to `{ onClick: foo }`. With this
    /// option it's compiled to:
    /// ```js
    /// { onClick: _cache[0] || (_cache[0] = e => _ctx.foo(e)) }
    /// ```
    /// - Requires "prefixIdentifiers" to be enabled because it relies on scope
    /// analysis to determine if a handler is safe to cache.
    /// @default false
    pub cache_handlers: bool,

    /// - `module` mode will generate ES module import statements for helpers
    /// and export the render function as the default export.
    /// - `function` mode will generate a single `const { helpers... } = Vue`
    /// statement and return the render function. It expects `Vue` to be globally
    /// available (or passed by wrapping the code with an IIFE). It is meant to be
    /// used with `new Function(code)()` to generate a render function at runtime.
    /// @default 'function'
    pub mode: ScriptMode,
    /// Generate source map?
    /// @default false
    pub source_map: bool,
    /// Whether the output JS needs re-rendering when Vue runtime data change.
    /// e.g. SSR can set it to false since SSR is executed only once per request.
    /// @default true
    pub need_reactivity: bool,
    /// Custom error reporter. Default is noop.
    pub error_handler: RcErrHandle,
    // deleted options
    // nodeTransforms?: NodeTransform[]
    // transformHoist?: HoistTransform | null
    // expressionPlugins?: ParserPlugin[]
    // prefix_identifiers: bool,
    // optimizeImports?: boolean // farewell, webpack optimization

    // moved to SFCInfo
    // bindingMetadata?: BindingMetadata
    // inline?: boolean
    // filename?: string
    // scopeId?: string | null
    // slotted?: boolean

    // moved to SSR or need_reactivity
    // ssr: bool // will be false in fallback node
    // inSSR?: bool // always true in ssr build
    // ssrCssVars?: string
    // ssrRuntimeModuleName?: string
}

impl Default for CompileOption {
    fn default() -> Self {
        let mut directive_converters = FxHashMap::default();
        directive_converters.insert(V_BIND.0, V_BIND.1);
        directive_converters.insert(V_MODEL.0, V_MODEL.1);
        directive_converters.insert("on", no_op_directive_convert);
        Self {
            is_native_tag: yes,
            is_void_tag: no,
            is_pre_tag: no,
            get_builtin_component: |_| None,
            is_custom_element: no,
            get_namespace: |_, _| Namespace::Html,
            get_text_mode: |_| TextMode::Data,
            delimiters: ("{{".into(), "}}".into()),
            whitespace: WhitespaceStrategy::Preserve,
            helper_strs: &[],
            preserve_comments: None,
            is_dev: true,
            directive_converters,
            hoist_static: false,
            cache_handlers: false,
            mode: ScriptMode::Function {
                prefix_identifier: false,
                runtime_global_name: "Vue".into(),
            },
            source_map: false,
            need_reactivity: true,
            error_handler: Rc::new(NoopErrorHandler),
        }
    }
}

impl CompileOption {
    pub fn scanning(&self) -> ScanOption {
        ScanOption {
            delimiters: self.delimiters.clone(),
            get_text_mode: self.get_text_mode,
        }
    }
    pub fn parsing(&self) -> ParseOption {
        ParseOption {
            whitespace: self.whitespace.clone(),
            preserve_comment: self.preserve_comments.unwrap_or(self.is_dev),
            get_namespace: self.get_namespace,
            get_text_mode: self.get_text_mode,
            is_native_element: self.is_native_tag,
            is_void_tag: self.is_void_tag,
            is_pre_tag: self.is_pre_tag,
            get_builtin_component: self.get_builtin_component,
            is_custom_element: self.is_custom_element,
        }
    }
    pub fn converting(&self) -> ConvertOption {
        ConvertOption {
            get_builtin_component: self.get_builtin_component,
            is_dev: self.is_dev,
            directive_converters: self.directive_converters.clone(),
            need_reactivity: self.need_reactivity,
        }
    }
    pub fn transforming(&self) -> TransformOption {
        let prefix = match self.mode {
            ScriptMode::Function {
                prefix_identifier, ..
            } => prefix_identifier,
            ScriptMode::Module { .. } => true,
        };
        TransformOption {
            prefix_identifier: prefix,
            is_dev: self.is_dev,
        }
    }
    pub fn codegen(&self) -> CodeGenerateOption {
        CodeGenerateOption {
            is_dev: self.is_dev,
            mode: self.mode.clone(),
            source_map: self.source_map,
            helper_strs: self.helper_strs,
        }
    }
}

// TODO: refactor this ownership usage
pub trait TemplateCompiler<'a> {
    type IR;
    type Info: Copy;
    type Output;

    fn scan(&self, source: &'a str) -> Tokens<'a>;
    fn parse(&self, tokens: Tokens<'a>) -> AstRoot<'a>;
    fn convert(&self, ast: AstRoot<'a>, info: Self::Info) -> Self::IR;
    fn transform(&self, ir: &mut Self::IR, info: Self::Info);
    fn generate(&self, ir: Self::IR, info: Self::Info) -> Self::Output;
    fn get_error_handler(&self) -> RcErrHandle;

    fn compile(&self, source: &'a str, info: Self::Info) -> Self::Output {
        let tokens = self.scan(source);
        let ast = self.parse(tokens);
        let mut ir = self.convert(ast, info);
        self.transform(&mut ir, info);
        self.generate(ir, info)
    }
}

pub struct BaseCompiler<'a, P, W>
where
    W: io::Write,
    P: CorePass<BaseInfo<'a>>,
{
    writer: fn() -> W,
    passes: fn(&'a SFCInfo<'a>, &CompileOption) -> P,
    option: CompileOption,
    scanner: Scanner,
    parser: Parser,
    pd: PhantomData<&'a ()>,
}

impl<'a, P, W> BaseCompiler<'a, P, W>
where
    W: io::Write,
    P: CorePass<BaseInfo<'a>>,
{
    pub fn new(
        writer: fn() -> W,
        passes: fn(&'a SFCInfo<'a>, &CompileOption) -> P,
        option: CompileOption,
    ) -> Self {
        Self {
            writer,
            passes,
            scanner: Scanner::new(option.scanning()),
            parser: Parser::new(option.parsing()),
            option,
            pd: PhantomData,
        }
    }
    fn get_converter(&self) -> BaseConverter {
        let eh = self.get_error_handler();
        let option = self.option.converting();
        BaseConverter::new(eh, option)
    }
}

impl<'a, P, W> TemplateCompiler<'a> for BaseCompiler<'a, P, W>
where
    W: io::Write,
    P: CorePass<BaseInfo<'a>>,
{
    type IR = BaseRoot<'a>;
    type Info = &'a SFCInfo<'a>;
    type Output = io::Result<W>;

    fn scan(&self, source: &'a str) -> Tokens<'a> {
        self.scanner.scan(source, self.get_error_handler())
    }

    fn parse(&self, tokens: Tokens<'a>) -> AstRoot<'a> {
        self.parser.parse(tokens, self.get_error_handler())
    }
    fn convert(&self, ast: AstRoot<'a>, info: Self::Info) -> Self::IR {
        self.get_converter().convert_ir(ast, info)
    }
    fn transform(&self, ir: &mut Self::IR, info: Self::Info) {
        let pass = (self.passes)(info, &self.option);
        BaseTransformer::transform(ir, pass)
    }
    fn generate(&self, ir: Self::IR, sfc_info: Self::Info) -> Self::Output {
        let mut writer = (self.writer)();
        let option = self.option.codegen();
        let generator = CodeGen::new(option);
        let gen_info = CodeGenInfo {
            writer: &mut writer,
            sfc_info,
        };
        generator.generate(ir, gen_info)?;
        Ok(writer)
    }
    fn get_error_handler(&self) -> RcErrHandle {
        self.option.error_handler.clone()
    }
}

pub fn get_base_passes<'a>(
    sfc_info: &'a SFCInfo<'a>,
    opt: &CompileOption,
) -> impl CorePass<BaseInfo<'a>> {
    use crate::chain;
    let prefix_identifier = opt.transforming().prefix_identifier;
    let shared = chain![
        SlotFlagMarker,
        ExpressionProcessor {
            prefix_identifier,
            sfc_info,
            err_handle: opt.error_handler.clone(),
        },
    ];
    chain![
        TextOptimizer,
        EntityCollector::default(),
        PatchFlagMarker,
        SharedInfoPasses {
            passes: shared,
            shared_info: Scope::default(),
            pd: PhantomData,
        },
        HoistStatic::default(),
    ]
}
