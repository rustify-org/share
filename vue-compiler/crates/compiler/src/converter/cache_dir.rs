// v-once / v-memo
use super::{BaseConversion, BaseIR, CoreConversion, Directive, Element};
use crate::ir::{IRNode, CacheIR, CacheKind, JsExpr as Js};
use crate::error::CompilationErrorKind as ErrorKind;
use crate::util::find_dir_empty;

pub fn pre_convert_memo<'a>(elem: &mut Element<'a>) -> Option<Directive<'a>> {
    let dir = find_dir_empty(&mut *elem, "memo")?;
    let b = dir.take();
    Some(b)
}

pub fn pre_convert_once<'a>(elem: &mut Element<'a>) -> Option<Directive<'a>> {
    let dir = find_dir_empty(&mut *elem, "once")?;
    let b = dir.take();
    // don't use cache if ancestor already in v-once/v-memo
    let children = elem.children.iter_mut().filter_map(|c| c.get_element_mut());
    for child in children {
        pre_convert_once(child);
    }
    Some(b)
}

pub fn convert_memo<'a>(bc: &BaseConversion, d: Directive<'a>, n: BaseIR<'a>) -> BaseIR<'a> {
    if let Some(error) = d.check_empty_expr(ErrorKind::VMemoNoExpression) {
        bc.emit_error(error);
        return n;
    }
    let mut n = n;
    // non-component sub tree should be turned into a block
    if let IRNode::VNodeCall(vnode) = &mut n {
        if !vnode.is_component {
            vnode.is_block = true;
        }
    }
    let expr_raw = d.expression.expect("v-memo should not be empty");
    let expr = Js::simple(expr_raw.content);
    IRNode::CacheNode(CacheIR {
        kind: CacheKind::Memo(expr),
        child: Box::new(n),
    })
}

pub fn convert_once<'a>(_bc: &BaseConversion, _: Directive<'a>, n: BaseIR<'a>) -> BaseIR<'a> {
    IRNode::CacheNode(CacheIR {
        kind: CacheKind::Once,
        child: Box::new(n),
    })
}

#[cfg(test)]
mod test {
    use super::super::test::base_convert;
    use super::*;
    use crate::cast;

    #[test]
    fn test_memo() {
        let case = "<p v-memo='a'/>";
        let mut body = base_convert(case).body;
        assert_eq!(body.len(), 1);
        let cn = cast!(body.remove(0), IRNode::CacheNode);
        let n = cast!(cn.kind, CacheKind::Memo);
        let expr = cast!(n, Js::Simple);
        assert_eq!(expr.into_string(), "a");
    }
    #[test]
    fn test_memo_in_v_if() {
        let case = "<p v-if='a' v-memo='a'/>";
        let mut body = base_convert(case).body;
        assert_eq!(body.len(), 1);
        let i = cast!(body.remove(0), IRNode::If);
        let child = &*i.branches[0].child;
        cast!(child, IRNode::CacheNode);
    }
    #[test]
    fn test_memo_in_v_for() {
        let case = "<p v-for='a in b' v-memo='a'/>";
        let mut body = base_convert(case).body;
        let f = cast!(body.remove(0), IRNode::For);
        let cn = cast!(*f.child, IRNode::CacheNode);
        assert!(matches!(cn.kind, CacheKind::MemoInVFor { .. }));
    }
    #[test]
    fn test_memo_in_template_for() {
        let case = "<template v-for='a in b'><p v-memo='a'/></template>";
        let mut body = base_convert(case).body;
        let f = cast!(body.remove(0), IRNode::For);
        let mut vn = cast!(*f.child, IRNode::VNodeCall);
        let cn = cast!(vn.children.remove(0), IRNode::CacheNode);
        cast!(cn.kind, CacheKind::Memo);
    }
    // fn test_once() {
    //     let cases = [
    //         "<template v-for='a in b'><p v-once/></template>",
    //         "<p v-for='a in b' v-once/>",
    //         "<p v-if='a' v-once/>",
    //         "<p v-once/>",
    //     ];
    // }
}
