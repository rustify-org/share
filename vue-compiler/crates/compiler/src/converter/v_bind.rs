use crate::error::CompilationErrorKind as ErrorKind;
use crate::flags::RuntimeHelper;

use super::{
    CoreDirConvRet, Directive, DirectiveConvertResult, DirectiveConverter, Element, ErrorHandler,
    JsExpr as Js,
};
use crate::parser::DirectiveArg;

// this module process v-bind without arg and with arg.
pub fn convert_v_bind<'a>(
    dir: &mut Directive<'a>,
    _: &Element<'a>,
    eh: &dyn ErrorHandler,
) -> CoreDirConvRet<'a> {
    let expr = if let Some(error) = dir.check_empty_expr(ErrorKind::VBindNoExpression) {
        eh.on_error(error);
        if dir.argument.is_none() {
            return DirectiveConvertResult::Dropped;
        } else {
            // <p :test> returns {test: ""}
            Js::str_lit("")
        }
    } else {
        let expr = dir
            .expression
            .take()
            .expect("dir without value should be dropped");
        Js::simple(expr.content)
    };
    let Directive {
        modifiers,
        argument,
        ..
    } = dir;
    let value = if let Some(arg) = argument {
        let mut arg = match arg {
            DirectiveArg::Static(s) => Js::str_lit(*s),
            DirectiveArg::Dynamic(s) => {
                let e = Js::simple(*s);
                Js::Compound(vec![Js::Src("("), e, Js::Src(") || ''")])
            }
        };
        // TODO: handle .attr, .prop, modifiers in DOM
        if modifiers.contains(&"camel") {
            arg = match arg {
                Js::StrLit(ref mut s) => {
                    s.camelize();
                    arg
                }
                a => Js::Call(RuntimeHelper::CAMELIZE, vec![a]),
            }
        }
        Js::Props(vec![(arg, expr)])
    } else {
        expr
    };
    DirectiveConvertResult::Converted {
        value,
        runtime: Err(false),
    }
}

pub const V_BIND: DirectiveConverter = ("bind", convert_v_bind);

#[cfg(test)]
mod test {}
