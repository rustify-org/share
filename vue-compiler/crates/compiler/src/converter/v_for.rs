use super::{
    BaseConvertInfo, BaseConversion, BaseIR, CompilationError, ConvertInfo, CoreConversion,
    Directive, Element,
};
use crate::{
    error::CompilationErrorKind as ErrorKind,
    flags::PatchFlag,
    ir::{ForNodeIR, ForParseResult, IRNode, JsExpr as Js},
    parser::ElementType,
    util::{find_dir_empty, find_prop, VStr},
};
use smallvec::SmallVec;

/// Pre converts v-if or v-for like structural dir
pub fn pre_convert_for<'a, T: ConvertInfo, C: CoreConversion<'a, T> + ?Sized>(
    bc: &C,
    elem: &mut Element<'a>,
) -> Option<Directive<'a>> {
    // convert v-for, v-if is converted elsewhere
    let dir = find_dir_empty(&mut *elem, "for")?;
    let b = dir.take();
    debug_assert!(find_dir_empty(&mut *elem, "for").is_none());
    check_template_v_for_key(bc, elem);
    Some(b)
}

pub fn convert_for<'a>(bc: &BaseConversion, d: Directive<'a>, n: BaseIR<'a>) -> BaseIR<'a> {
    // on empty v-for expr error
    if let Some(error) = d.check_empty_expr(ErrorKind::VForNoExpression) {
        bc.emit_error(error);
        return n;
    }
    let expr = d.expression.expect("v-for must have expression");
    let (source, parse_result) = match parse_for_expr(expr.content) {
        Some(parsed) => parsed,
        None => {
            let error = CompilationError::new(ErrorKind::VForMalformedExpression)
                .with_location(expr.location.clone());
            bc.emit_error(error);
            return n;
        }
    };
    let n = convert_memo_in_v_for(n, || parse_result.key.clone());
    IRNode::For(ForNodeIR {
        source,
        parse_result,
        child: Box::new(n),
        key: None,
        is_stable: false,
        fragment_flag: PatchFlag::empty(),
    })
}

type ParsedFor<'a> = (Js<'a>, ForParseResult<BaseConvertInfo<'a>>);

const PARENS: &[char] = &['(', ')'];
fn parse_for_expr(expr: VStr) -> Option<ParsedFor> {
    // split source and binding
    let (lhs, rhs) = expr
        .raw
        .split_once(" in ")
        .or_else(|| expr.raw.split_once(" of "))
        .map(|(l, r)| (l.trim_matches(PARENS), r.trim()))?;
    if rhs.is_empty() {
        return None;
    }
    // split iterator by ,
    let (val, key, idx) = split_v_for_iter(lhs);
    Some((
        Js::simple(rhs.trim()),
        ForParseResult {
            value: Js::Param(val),
            key: key.map(Js::Param),
            index: idx.map(Js::Param),
        },
    ))
}

const DESTRUCTING: &[char] = &['}', ']'];
fn split_v_for_iter(mut lhs: &str) -> (&str, Option<&str>, Option<&str>) {
    let mut split = SmallVec::<[&str; 3]>::new();
    while let Some((pre, post)) = lhs.rsplit_once(',') {
        if post.contains(DESTRUCTING) || split.len() == 2 {
            break;
        }
        lhs = pre;
        split.push(post.trim());
    }
    split.push(lhs.trim());
    split.reverse();
    match split.len() {
        2 => (split[0], Some(split[1]), None),
        3 => (split[0], Some(split[1]), Some(split[2])),
        _ => (split[0], None, None),
    }
}

fn convert_memo_in_v_for<'a, F>(n: BaseIR<'a>, get_key: F) -> BaseIR<'a>
where
    F: Fn() -> Option<Js<'a>>,
{
    use crate::ir::{
        CacheIR,
        CacheKind::{Memo, MemoInVFor},
    };
    use IRNode::CacheNode;
    let (expr, child) = match n {
        CacheNode(CacheIR {
            kind: Memo(expr),
            child,
        }) => (expr, child),
        _ => return n,
    };
    let kind = MemoInVFor {
        v_for_key: get_key(),
        expr,
    };
    CacheNode(CacheIR { kind, child })
}

// check <template v-for> key placement
fn check_template_v_for_key<'a, T: ConvertInfo, C: CoreConversion<'a, T> + ?Sized>(
    bc: &C,
    elem: &Element,
) {
    if elem.tag_type != ElementType::Template {
        return;
    }
    let first_wrong = elem
        .children
        .iter()
        .filter_map(|child| child.get_element())
        .find_map(|child| find_prop(child, "key"));
    if let Some(wrong) = first_wrong {
        let key_loc = wrong.get_ref().get_location().clone();
        let error =
            CompilationError::new(ErrorKind::VForTemplateKeyPlacement).with_location(key_loc);
        bc.emit_error(error);
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::cast;
    fn to_str(e: Js) -> &str {
        cast!(e, Js::Param)
    }
    fn check_equal(src: &str, expect: (&str, &str, Option<&str>, Option<&str>)) {
        let (src, ret) = parse_for_expr(VStr::raw(src)).expect("should parse");
        let source = cast!(src, Js::Simple);
        assert_eq!(source.into_string(), expect.0);
        assert_eq!(to_str(ret.value), expect.1);
        assert_eq!(ret.key.map(to_str), expect.2);
        assert_eq!(ret.index.map(to_str), expect.3);
    }
    #[test]
    fn test_parse_for_expr() {
        for &(src, expect) in &[
            ("a in [123]", ("[123]", "a", None, None)),
            ("   in [123]", ("[123]", "", None, None)),
            ("   a      in     [123]    ", ("[123]", "a", None, None)),
            ("a, b, c   in p ", ("p", "a", "b".into(), "c".into())),
            ("{a, b, c} in p ", ("p", "{a, b, c}", None, None)),
            ("{a, b}, c in p ", ("p", "{a, b}", "c".into(), None)),
            ("[a,] , b in p ", ("p", "[a,]", "b".into(), None)),
            ("a,b,c,d,e in p ", ("p", "a,b,c", "d".into(), "e".into())),
            ("(a,b) in p ", ("p", "a", "b".into(), None)),
            ("(a,b, c, d) in p ", ("p", "a,b", "c".into(), "d".into())),
            ("(,,,) in p ", ("p", ",", "".into(), "".into())),
            ("(,,) in p ", ("p", "", "".into(), "".into())),
        ] {
            check_equal(src, expect);
        }
    }

    #[test]
    fn test_parse_invalid_for() {
        for src in &["", "           in             "] {
            assert!(parse_for_expr(VStr::raw(src)).is_none());
        }
    }
}
