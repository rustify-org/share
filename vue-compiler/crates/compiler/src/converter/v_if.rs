use super::{
    super::parser::ElemProp, AstNode, BaseConvertInfo, BaseConversion as BC, BaseIR,
    CompilationError, Directive, Element, IRNode,
};
use crate::{
    converter::{CoreConversion, JsExpr as Js},
    error::CompilationErrorKind as ErrorKind,
    ir::{IfBranch, IfNodeIR},
    scanner::Attribute,
    util::{find_dir_empty, find_prop, VStr},
};
use rustc_hash::FxHashSet;
use std::{iter::Peekable, vec::IntoIter};

// TODO: reduce vec allocation by using Drain iter
// but using drain need GAT
pub enum PreGroup<'a> {
    VIfGroup(Vec<Element<'a>>),
    StandAlone(AstNode<'a>),
}

struct PreGroupIter<'a> {
    inner: Peekable<IntoIter<AstNode<'a>>>,
    group: Vec<Element<'a>>,
}

impl<'a> PreGroupIter<'a> {
    fn new(children: Vec<AstNode<'a>>) -> Self {
        let len = children.len();
        Self {
            inner: children.into_iter().peekable(),
            group: Vec::with_capacity(len),
        }
    }
    fn flush_group(&mut self) -> Option<PreGroup<'a>> {
        if self.group.is_empty() {
            None
        } else {
            let group = self.group.drain(..).collect();
            Some(PreGroup::VIfGroup(group))
        }
    }

    fn next_standalone(&mut self) -> Option<PreGroup<'a>> {
        // either iter is empty or has no if/else/else-if
        debug_assert!(self
            .inner
            .peek()
            .and_then(|n| n.get_element())
            .and_then(|e| find_dir_empty(e, ["if", "else", "else-if"]))
            .is_none());
        self.inner.next().map(PreGroup::StandAlone)
    }
}
impl<'a> Iterator for PreGroupIter<'a> {
    type Item = PreGroup<'a>;
    fn next(&mut self) -> Option<Self::Item> {
        while let Some(n) = self.inner.peek() {
            // group elements if they have v-if/v-else
            let found = n
                .get_element()
                .and_then(|e| find_dir_empty(e, ["if", "else-if", "else"]));
            if let Some(d) = found {
                // separate v-if into different groups
                if d.get_ref().name == "if" && !self.group.is_empty() {
                    return self.flush_group();
                }
                let n = self.inner.next().unwrap(); // must next to advance
                self.group.push(n.into_element());
            } else if let AstNode::Text(s) = n {
                if self.group.is_empty() || !s.is_all_whitespace() {
                    // break if text is not whitespaces
                    // or no preceding v-if
                    break;
                }
                // skip whitespace when v-if precedes
                self.inner.next().unwrap();
            } else if matches!(n, &AstNode::Comment(_)) {
                // ignore comments for now. #3619
                return self.next_standalone();
            } else {
                break;
            }
        }
        // vec emptied or next element has no v-if
        // first, flush preceding group
        self.flush_group().or_else(|| {
            // if no group, consume next standalone element if available
            self.next_standalone()
        })
    }
    #[inline]
    fn size_hint(&self) -> (usize, Option<usize>) {
        (0, self.inner.size_hint().1)
    }
}

/// pre group adjacent elements with v-if
// using generator here will be super cool
pub fn pre_group_v_if(children: Vec<AstNode>) -> impl Iterator<Item = PreGroup> {
    PreGroupIter::new(children)
}

/// key is Vue-generated default key based on the number of sibling v-if.
pub fn convert_if<'a>(c: &BC<'a>, elems: Vec<Element<'a>>, key: usize) -> BaseIR<'a> {
    debug_assert!(!elems.is_empty());
    check_v_if_group(c, &elems);
    let branches: Vec<_> = elems
        .into_iter()
        .enumerate()
        .map(|(i, n)| convert_if_branch(c, n, key + i))
        .collect();
    IRNode::If(IfNodeIR { branches })
}

pub fn report_dangling_else<'a>(c: &BC<'a>, elem: &Element<'a>) {
    debug_assert!(find_dir_empty(elem, "if").is_none());
    let loc = find_dir_empty(elem, ["else-if", "else"])
        .expect("must have other v-if dir")
        .get_ref()
        .location
        .clone();
    let error = CompilationError::new(ErrorKind::VElseNoAdjacentIf).with_location(loc);
    c.emit_error(error);
}

fn check_duplicate_key<'a>(
    c: &BC<'a>,
    prop: &ElemProp<'a>,
    dirs: &mut FxHashSet<VStr<'a>>,
    attrs: &mut FxHashSet<VStr<'a>>,
) {
    match prop {
        ElemProp::Dir(Directive {
            expression: Some(v),
            ..
        }) => {
            if dirs.contains(&v.content) {
                let error =
                    CompilationError::new(ErrorKind::VIfSameKey).with_location(v.location.clone());
                c.emit_error(error);
            } else {
                dirs.insert(v.content);
            }
        }
        ElemProp::Attr(Attribute { value: Some(v), .. }) => {
            if attrs.contains(&v.content) {
                let error =
                    CompilationError::new(ErrorKind::VIfSameKey).with_location(v.location.clone());
                c.emit_error(error);
            } else {
                attrs.insert(v.content);
            }
        }
        _ => (),
    }
}

fn check_v_if_group<'a>(c: &BC<'a>, elems: &[Element<'a>]) {
    // 1. check dangling else
    if find_dir_empty(&elems[0], "if").is_none() {
        report_dangling_else(c, &elems[0]);
    }
    if !c.option.is_dev {
        return;
    }
    // 2. check duplicate v-if key in dev build
    let mut dirs = FxHashSet::default();
    let mut attrs = FxHashSet::default();
    let mut has_else = false;
    for child in elems {
        if has_else {
            report_dangling_else(c, child);
            continue;
        }
        let prop = find_prop(child, "key");
        if let Some(prop) = prop {
            check_duplicate_key(c, prop.get_ref(), &mut dirs, &mut attrs);
        }
        has_else = find_dir_empty(child, ["else"]).is_some();
    }
}

fn convert_if_branch<'a>(
    c: &BC<'a>,
    mut e: Element<'a>,
    key: usize,
) -> IfBranch<BaseConvertInfo<'a>> {
    let dir = find_dir_empty(&mut e, ["if", "else-if", "else"])
        .expect("the element must have v-if directives")
        .take();
    report_duplicate_v_if(c, &mut e);
    let condition = convert_if_condition(c, dir);
    IfBranch {
        child: Box::new(c.pre_convert_element(e)),
        condition,
        info: key,
    }
}
fn convert_if_condition<'a>(c: &BC<'a>, dir: Directive<'a>) -> Option<Js<'a>> {
    if dir.name != "else" {
        if let Some(err) = dir.check_empty_expr(ErrorKind::VIfNoExpression) {
            c.emit_error(err);
            return Some(Js::Src("true"));
        }
    } else if let Some(expr) = dir.expression {
        let error =
            CompilationError::new(ErrorKind::UnexpectedDirExpression).with_location(expr.location);
        c.emit_error(error);
        return None;
    }
    dir.expression.map(|v| Js::simple(v.content))
}
fn report_duplicate_v_if<'a>(c: &BC<'a>, e: &mut Element<'a>) {
    // https://stackoverflow.com/a/48144226/2198656
    while let Some(found) = find_dir_empty(&mut *e, ["if", "else-if", "else"]) {
        let dir = found.take();
        let error = CompilationError::new(ErrorKind::VIfDuplicateDir).with_location(dir.location);
        c.emit_error(error);
    }
}

#[cfg(test)]
mod test {
    use super::super::test::*;
    use super::*;
    use crate::cast;

    fn test_no_panic() {
        let cases = [
            r#"
<p v-if="false">a</p>
<p v-else v-if="true">b</p>
<p v-else>c</p>"#,
            r#"<p v-if="123"/><p v-else="33"/>"#,
            r#"<p v-if/>"#,
            r#"<p v-if="1"/><p v-else-if="2"/><comp v-else/>"#, // key = 1, 2, 3
        ];
        for case in cases {
            base_convert(case);
        }
    }

    #[test]
    fn test_v_if() {
        let body = base_convert("<p v-if='true'/>").body;
        assert_eq!(body.len(), 1);
        let v_if = cast!(&body[0], IRNode::If);
        assert_eq!(v_if.branches.len(), 1);
        let condition = v_if.branches[0].condition.as_ref().unwrap();
        let cond = cast!(condition, Js::Simple);
        assert_eq!(cond.into_string(), "true");
    }
}
