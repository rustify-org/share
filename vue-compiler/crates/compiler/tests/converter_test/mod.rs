use super::common::serialize_yaml;
use super::common::TestErrorHandler;
use super::parser_test::base_parse;
use compiler::SFCInfo;
use compiler::converter::{self as C, BaseConverter, ConvertOption, Converter};
use crate::meta_macro;
use vue_compiler_core as compiler;

fn assert_ir(case: &str) -> String {
    let opt = SFCInfo::default();
    let ir = base_convert(case, &opt);
    serialize_yaml(ir.body)
}

meta_macro!(assert_ir);

#[test]
fn test_text_call() {
    assert_ir![["hello world", "hello {{world}}", "hello < world"]];
}

pub fn base_convert<'a>(s: &'a str, opt: &'a SFCInfo<'a>) -> C::BaseRoot<'a> {
    let ast = base_parse(s);
    let converter =
        BaseConverter::new(std::rc::Rc::new(TestErrorHandler), ConvertOption::default());
    converter.convert_ir(ast, opt)
}
