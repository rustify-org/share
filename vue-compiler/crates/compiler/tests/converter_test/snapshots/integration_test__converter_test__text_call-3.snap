---
source: crates/compiler/tests/converter_test/mod.rs
assertion_line: 23
expression: hello < world
---
- TextCall:
    fast_path: false
    need_patch: false
    texts:
      - StrLit: "hello "
      - StrLit: "<"
      - StrLit: " world"

